server/src/main/java/org/elasticsearch/index/shard/GlobalCheckpointListeners.java

```diff
@@ -53,6 +53,16 @@
      */
     @FunctionalInterface
     public interface GlobalCheckpointListener {
+
+        /**
+         * The executor on which the listener is notified. Defaults to the calling thread.
+         *
+         * @return the executor
+         */
+        default Executor executor() {
+            return Runnable::run;
```

Comment:
It seems this default is only used in tests (and the interface is only implemented once anyways), should we maybe just extract some utility method into tests to build these listeners?

E.g.

```java
    public static GlobalCheckpointListeners.GlobalCheckpointListener syncListener(BiConsumer<Long, Exception> consumer) {
        return new GlobalCheckpointListeners.GlobalCheckpointListener() {
            @Override
            public Executor executor() {
                return Runnable::run;
            }

            @Override
            public void accept(long globalCheckpoint, Exception e) {
                consumer.accept(globalCheckpoint, e);
            }
        };
    }
```

It seems a little strange to mark this as a `@FunctionalInterface` when it really only is one in tests?

Reply:
I pushed b565820a862107f74bfdd23a99158a820e1231f9. I mistakenly combined your two suggestions into a single commit, sorry.